File: HyperNovaLabsBackend/HyperNovaLabs.Api/Services/EmployeeService.cs
namespace HyperNovaLabs.Api.Services
{
    using Google.Protobuf.WellKnownTypes;

    using Grpc.Core;

    using HyperNovaLabs.Api.Protos.Models;
    using HyperNovaLabs.Api.Protos.Services;

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;

    public class EmployeeService : Protos.Services.EmployeeService.EmployeeServiceBase
    {
        private readonly Models.ExpenseReportContext Database;

        public EmployeeService(Models.ExpenseReportContext Context)
        {
            Database = Context;
        }

        public override Task<ListEmployeeResponse> ListEmployee(Empty Request, ServerCallContext Context)
        {
            ListEmployeeResponse Response = new();

            var Data = Database.Employees.Select(E => new Employee
            {
                Name = E.Name,
                Identification = E.Identification,
                Department = E.Department,
                Position = E.Position,
                Supervisor = E.Supervisor
            });

            Response.Employees.AddRange(Data);

            return Task.FromResult(Response);
        }

        public override async Task<Employee> GetEmployee(GetEmployeeRequest Request, ServerCallContext Context)
        {
            var Employee = await Database.Employees.FindAsync(Request.Identification);

            if (Employee is not null)
            {
                return new Employee
                {
                    Name = Employee.Name,
                    Identification = Employee.Identification,
                    Department = Employee.Department,
                    Position = Employee.Position,
                    Supervisor = Employee.Supervisor
                };
            }

            return default;
        }

        public override async Task<CreateEmployeeResponse> CreateEmployee(CreateEmployeeRequest Request, ServerCallContext Context)
        {
            CreateEmployeeResponse Response = new()
            {
                Employee = Request.Employee
            };

            try
            {
                var Employee = await Database.Employees.FindAsync(Request.Employee.Identification);

                if (Employee is not null)
                {
                    Response.HasError = true;
                    Response.Errors.Add("Un empleado con el mismo número de identificación ya esta registrado.");
                    return Response;
                }

                await Database.Employees.AddAsync(new Models.Employee
                {
                    Name = Request.Employee.Name,
                    Identification = Request.Employee.Identification,
                    Department = Request.Employee.Department,
                    Position = Request.Employee.Position,
                    Supervisor = Request.Employee.Supervisor
                });

                await Database.SaveChangesAsync();
            }
            catch (Exception Ex)
            {
                Response.HasError = true;

                while (Ex != null)
                {
                    Response.Errors.Add(Ex.Message);
                    Ex = Ex.InnerException;
                }
            }

            return Response;
        }

        public override async Task<UpdateEmployeeResponse> UpdateEmployee(UpdateEmployeeRequest Request, ServerCallContext Context)
        {
            UpdateEmployeeResponse Response = new()
            {
                Employee = Request.Employee
            };

            try
            {
                var Employee = await Database.Employees.FindAsync(Request.Employee.Identification);

                if (Employee is not null)
                {
                    Employee.Name = Request.Employee.Name;
                    Employee.Department = Request.Employee.Department;
                    Employee.Position = Request.Employee.Position;
                    Employee.Supervisor = Request.Employee.Supervisor;

                    Database.Employees.Update(Employee);
                    await Database.SaveChangesAsync();

                    Response.Success = true;
                    Response.Employee.Name = Employee.Name;
                    Response.Employee.Department = Employee.Department;
                    Response.Employee.Position = Employee.Position;
                    Response.Employee.Supervisor = Employee.Supervisor;
                }
                else
                {
                    Response.Errors.Add($"El empleado con el número de identificación \"{Request.Employee.Identification}\" no está registrado.");
                }
            }
            catch (Exception Ex)
            {
                while (Ex != null)
                {
                    Response.Errors.Add(Ex.Message);
                    Ex = Ex.InnerException;
                }
            }

            return Response;
        }

        public override async Task<DeleteEmployeeResponse> DeleteEmployee(DeleteEmployeeRequest Request, ServerCallContext Context)
        {
            DeleteEmployeeResponse Response = new();

            try
            {
                var Employee = await Database.Employees.FindAsync(Request.Identification);

                if (Employee is not null)
                {
                    Database.Employees.Remove(Employee);
                    await Database.SaveChangesAsync();

                    Response.Success = true;
                    Response.Employee = new Employee
                    {
                        Name = Employee.Name,
                        Identification = Employee.Identification,
                        Department = Employee.Department,
                        Position = Employee.Position,
                        Supervisor = Employee.Supervisor
                    };
                }
                else
                {
                    Response.Errors.Add($"El empleado con el número de identificación \"{Request.Identification}\" no está registrado.");
                }
            }
            catch (Exception Ex)
            {
                while (Ex != null)
                {
                    Response.Errors.Add(Ex.Message);
                    Ex = Ex.InnerException;
                }
            }

            return Response;
        }
    }
}

File: HyperNovaLabsNativeBackend/Backend/Sources/EmployeeService.cpp
#include "EmployeeService.hpp"

#include <iostream>
#include <memory>
#include <string>
#include <string_view>
#include <algorithm>

#include <google/protobuf/arena.h>

using namespace grpc;
using namespace google::protobuf;
using namespace HyperNovaLabs::Api::Protos::Services;
using namespace HyperNovaLabs::Api::Protos::Models;

Status EmployeeServiceImpl::ListEmployee(ServerContext *context, const Empty *request, ListEmployeeResponse *response)
{
  std::cout << "Ejecutando: ListEmployee\n";

  for (const auto &value : m_Store)
  {
    auto new_employee = response->add_employees();
    value.CopyTo(new_employee);
  }

  std::cout << "Finalizando: ListEmployee\n";

  return Status::OK;
}

Status EmployeeServiceImpl::GetEmployee(ServerContext *context, const GetEmployeeRequest *request, Employee *response)
{
  std::cout << "Ejecutando: GetEmployee\n";

  auto result = m_Store.Find(request->identification());

  if (result.has_value())
  {
    response->CopyFrom(*result);
  }

  std::cout << "Finalizando: GetEmployee\n";

  return Status::OK;
}

Status EmployeeServiceImpl::CreateEmployee(ServerContext *context, const CreateEmployeeRequest *request, CreateEmployeeResponse *response)
{
  static int64_t n_invokes = 0;

  std::cout << "Ejecutando: CreateEmployee (" << ++n_invokes << ")" << std::endl;

  m_Store.Add(request->employee());
  response->mutable_employee()->CopyFrom(request->employee());

  std::cout << "Finalizando: CreateEmployee\n";
  return Status::OK;
}

Status EmployeeServiceImpl::UpdateEmployee(ServerContext *context, const UpdateEmployeeRequest *request, UpdateEmployeeResponse *response)
{
  return Status::OK;
}

Status EmployeeServiceImpl::DeleteEmployee(ServerContext *context, const DeleteEmployeeRequest *request, DeleteEmployeeResponse *response)
{
  return Status::OK;
}

Status EmployeeServiceImpl::ListOfAnEmployeeBills(ServerContext *context, const ListOfAnEmployeesBillRequest *request, ListOfAnEmployeesBillResponse *response)
{
  return Status::OK;
}
